### ISSUE TITLE
Does .NET Core 3.0 on Linux support GCLargePages

### ISSUE THREAD

**Issue by tmds** [dotnet/coreclr#26696]
I'm trying to figure out if this is supported with .NET Core 3.0.

I see the code is there:

https://github.com/dotnet/coreclr/blob/922429db0144dd6f3b4324805464dae82857512a/src/pal/src/map/virtual.cpp#L991-L1002

When I (using 3.0.100-preview9-014004 on a `dotnet new web`):
```
$ export COMPlus_GCLargePages=1
$ strace -f -e mmap dotnet bin/Debug/netcoreapp3.0/web.dll
```
The app doesn't fail, and none of the `mmap` syscalls have the `MAP_HUGETLB` flag.

My system isn't configured for huge pages. I assume Microsoft preview9 build includes the feature, which is why I thought it would show up in the `strace`.

cc @janvorli

**Comment by tmds** [comment]
cc @omajid

**Comment by jkotas** [comment]
cc @Maoni0 @mjsabby 

The code for it is there (https://github.com/dotnet/coreclr/pull/24098), but it does not have any tests so it can very well be broken in various ways. I do not think we have declared the large pages as officially supported for .NET Core 3.0.

**Comment by tmds** [comment]
It's mentioned in the [What's new in .NET Core 3.0](https://docs.microsoft.com/en-us/dotnet/core/whats-new/dotnet-core-3-0), but I hadn't noticed this:

> on Windows

So maybe not supported on Linux then.

**Comment by Maoni0** [comment]
are you running in a container with a memory limit? this is just so that hardlimit is set. otherwise you'd need to set COMPlus_GCHeapHardLimit to set a hardlimit first before large pages can be enabled. 

I'm been meaning to write about large page support but haven't gotten around to it.

**Comment by mjsabby** [comment]
@tmds You need to do what @Maoni0 said which is set `COMPlus_GCHeapHardLimit` in addition to `COMPlus_GCLagePages`

This work item in the docs repo needs to be complete so we can inform folks what is required for Large Pages to work on various OSes:

https://github.com/dotnet/docs/issues/11931

@Maoni0 maybe we can sit next week and author this on the docs page? Ideally before 23rd Sep since that is RTM.

**Comment by tmds** [comment]
> set COMPlus_GCHeapHardLimit in addition to COMPlus_GCLagePages

Is it also needed when not running in a container (which is what I'm doing)?
Why is it needed?

**Comment by mjsabby** [comment]
Large Pages is only supported when the heap is put under hard limit. Moani was pointing that if you're running inside a container this gets set and so you potentially don't need to set it. Outside of a container you need to set it.

Large Pages are implemented differently on different OSes, so while on Linux you can reserve these pages at startup (or near startup), on Windows there is no such feature at the moment so you have to either commit all the memory you want to be in large pages upfront or you take your chance and may not be able to fulfill the large pages request at a later time in your app lifetime due to OS fragmentation. That sort of then raises another question, should you throw out of memory if you can't satisfy a large page allocation or should you fall back to small pages?

To avoid those issues in the 3.0 timeframe we chose that large pages only work when the heap hard limit is set.

**Comment by tmds** [comment]
> Large Pages is only supported when the heap is put under hard limit. Moani was pointing that if you're running inside a container this gets set and so you potentially don't need to set it. Outside of a container you need to set it.

ok! When I set `export COMPlus_GCHeapHardLimit=$((1024*1024*1024))`, the app crashes in an expected way (since I don't have large pages enabled):
```
[pid 13447] mmap(NULL, 137438953472, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_HUGETLB, -1, 0) = -1 ENOMEM (Cannot allocate memory)
[pid 13447] mmap(NULL, 68719476736, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_HUGETLB, -1, 0) = -1 ENOMEM (Cannot allocate memory)
Failed to create CoreCLR, HRESULT: 0x8007000E
```

> To avoid those issues in the 3.0 timeframe we chose that large pages only work when the heap hard limit is set.

This means that, on Linux, coreclr reserves `GCHeapHardLimit` of large pages when the app starts, to avoid allocation issues later (due to fragmentation)?

Is Large pages supported on Linux for 3.0? Or considered experimental?

**Comment by tmds** [comment]
> This means that, on Linux, coreclr reserves GCHeapHardLimit of large pages when the app starts, to avoid allocation issues later (due to fragmentation)?

Yes. Got my answer by [watching the video](https://channel9.msdn.com/Shows/On-NET/Walking-Down-the-Memory-Lane-with-Bing-Part-1) :)

**Comment by tmds** [comment]
```
$ dotnet --version
Failed to create CoreCLR, HRESULT: 0x8007000E
```
-> it's not clear to the user that this is due to large page allocation failure.

**Comment by tmds** [comment]
@mjsabby @Maoni0 can you confirm this is a supported feature on Linux for 3.0? Or should this be considered experimental?

**Comment by Maoni0** [comment]
@mjsabby I've been OOF. yes let's work on the doc for this.
@tmds I would consider this a supported feature on Windows; on Linux it should work but we haven't done nearly as much testing there so I would consider it experimental there.

**Comment by tmds** [comment]
Thanks you for clarifying.
